File: src/main/java/com/wesker926/leetcode/algorithms/p0071/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2022/1/6
// @description: 逐字符检查
package s1

func simplifyPath(path string) string {
	path += "/"
	res := make([]rune, 0, len(path))
	res = append(res, '/')
	for i, last, cur := 0, 0, 0; i < len(path); i++ {
		if path[i] != '/' {
			cur++
			res = append(res, rune(path[i]))
			continue
		}

		last = len(res) - 1
		if res[last] == '/' {
			continue
		}

		if res[last] != '.' || (cur == 2 && res[last-1] != '.') || cur > 2 {
			res = append(res, '/')
		} else if cur == 1 {
			res = res[:last]
		} else {
			res = res[:last-2]
			for last = len(res) - 1; last >= 0 && res[last] != '/'; last-- {
			}
			if last < 0 {
				last = 0
			}
			res = res[:last+1]
		}
		cur = 0
	}
	if len(res) != 1 && res[len(res)-1] == '/' {
		res = res[:len(res)-1]
	}
	return string(res)
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0071/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0071.s1;

/**
 * @author wesker.gh
 * @date 2022/5/3
 * @description 栈
 */
public class Solution {
    public String simplifyPath(String path) {
        char[] origin = (path + '/').toCharArray();
        char[] stack = new char[path.length() + 1];
        int idx = 0;
        stack[idx++] = '/';
        for (int i = 0, c = 0; i < origin.length; i++) {
            if (origin[i] != '/') {
                c++;
                stack[idx++] = origin[i];
                continue;
            }

            if (stack[idx - 1] == '/') {
                continue;
            }

            if (stack[idx - 1] != '.' || (c == 2 && stack[idx - 2] != '.') || c > 2) {
                stack[idx++] = '/';
            } else if (c == 1) {
                idx--;
            } else {
                idx -= 3;
                while (idx > 0 && stack[idx - 1] != '/') {
                    idx--;
                }
                if (idx == 0) {
                    idx++;
                }
            }
            c = 0;
        }
        if (idx != 1 && stack[idx - 1] == '/') {
            idx--;
        }
        return new String(stack, 0, idx);
    }
}
